tensorly/tenalg/proximal.py

```diff
@@ -248,3 +248,207 @@ def hals_nnls(UtM, UtU, V=None, n_iter_max=500, tol=10e-8,
+                        else:
+                            x_vec = tl.index_update(x_vec, tl.index[int(i)], 0)
+                    x_vec = tl.clip(x_vec, 0, tl.max(x_vec))
+                except:
+                    x_vec = tl.copy(x_vec)
+            break
+
+        if True in R:
+            indice = tl.argmax(w)
+            P = tl.index_update(P, tl.index[indice], True)
+            R = tl.index_update(R, tl.index[indice], False)
+        # To avoid singularity error when initial x exists
+        try:
+            a = tl.tensor(tl.to_numpy(AtA)[P, :][:, P], **tl.context(AtA))
+            b = AtB[P]
+            c = tl.solve(a, b)
+            indice_list = []
+            for i in range(tl.shape(s)[0]):
+                if P[i]:
+                    indice_list.append(i)
+                    s = tl.index_update(s, tl.index[int(i)], c[len(indice_list)-1])
+                else:
+                    s = tl.index_update(s, tl.index[int(i)], 0)
+        # Start from zeros if solve is not achieved  
+        except:
+            x_vec = tl.zeros(tl.shape(AtA)[1])
+            s = tl.zeros(tl.shape(x_vec), **tl.context(x_vec))
+            P = x_vec > 0
+            R = x_vec <= 0
+            if True in R:
+                indice = tl.argmax(w)
+                P = tl.index_update(P, tl.index[indice], True)
+                R = tl.index_update(R, tl.index[indice], False)
+            a = tl.tensor(tl.to_numpy(AtA)[P, :][:, P], **tl.context(AtA))
+            b = AtB[P]
+            c = tl.solve(a, b)
+            indice_list = []
+            for i in range(tl.shape(s)[0]):
+                if P[i]:
+                    indice_list.append(i)
+                    s = tl.index_update(s, tl.index[int(i)], c[len(indice_list)-1])
+                else:
+                    s = tl.index_update(s, tl.index[int(i)], 0)
+
+        # update support vector if it is necessary
+        if tl.min(s[P]) <= 0:
+            for i in range(len(P)):
+                alpha = tl.min(x_vec[P][s[P] <= 0] / (x_vec[P][s[P] <= 0] - s[P][s[P] <= 0]))
+                update = alpha * (s - x_vec)
+                x_vec = x_vec + update
+                P = x_vec > 0
+                R = x_vec <= 0                 
+                a = tl.tensor(tl.to_numpy(AtA)[P, :][:, P], **tl.context(AtA))
+                b = AtB[P]
+                c = tl.solve(a, b)
+                indice_list = []
+                for i in range(tl.shape(s)[0]):
+                    if P[i]:
+                        indice_list.append(i)
+                        s = tl.index_update(s, tl.index[int(i)], c[len(indice_list) - 1])
+                    else:
+                        s = tl.index_update(s, tl.index[int(i)], 0)
+
+                if True not in P or tl.min(s[P]) > 0:
+                    break
+        # set x to s
+        x_vec = tl.clip(s, 0, tl.max(s))
+
+        # gradient update
+        second = tl.dot(AtA, x_vec)
+        w = first - second
```

Comment:
same as above, no need to introduce second as it's not used anywhere else, just 
`AtB - tl.dot(AtA, x_vec)`

Reply:
There is no second or first anymore